File: solutions/2707. Extra Characters in a String/2707.py
# Time complexity: O(n^2)
# Space complexity: O(n)
class Solution:
  def minExtraChar(self, s: str, dictionary: List[str]) -> int:
    n = len(s)
    dictionarySet = set(dictionary)
    dp = [0] + [n] * n
        
    for i in range(1, n + 1):
      for j in range(i):
        if s[j:i] in dictionarySet:
          dp[i] = min(dp[i], dp[j])
        else:
          dp[i] = min(dp[i], dp[j] + i - j)
        
    return dp[n]

File: solutions/2707. Extra Characters in a String/2707.rs
// Time complexity: O(n^2)
// Space complexity: O(n)
use std::collections::HashSet;

impl Solution {
    pub fn min_extra_char(s: String, dictionary: Vec<String>) -> i32 {
        let n = s.len();
        let dictionary_set: HashSet<String> = dictionary.into_iter().collect();
        let mut dp = vec![n as i32; n + 1];
        dp[0] = 0;

        for i in 1..=n {
            for j in 0..i {
                let substring = &s[j..i];
                if dictionary_set.contains(substring) {
                    dp[i] = dp[i].min(dp[j]);
                } else {
                    dp[i] = dp[i].min(dp[j] + (i - j) as i32);
                }
            }
        }

        dp[n]
    }
}
